### ISSUE TITLE
Debian target compile error

### ISSUE THREAD

**Issue by akemnade** [koreader/koreader#9530]
* KOReader version: v2022.08-19-gc3cd88e0
* Device: Cubietruck with devuan chimeara armhf

#### Issue
compile failure due to unrecognized flag
```
g++ -I/data/src/koreader/koreader/base/thirdparty/kpvcrlib/crengine/crengine/include/ -L/data/src/koreader/koreader/base/build/arm-linux-gnueabihf/libs -O2 -ffast-math -pipe -fomit-frame-pointer   -mtune=generic -fPIC -I/data/src/koreader/koreader/base/thirdparty/luajit/build/arm-linux-gnueabihf/luajit-prefix/src/luajit/src -shared -Wl,-E -Wl,-rpath,'$ORIGIN' \
        -DLDOM_USE_OWN_MEM_MAN=1 -DUSE_SRELL_REGEX=1 \
         -fvisibility=hidden -o build/arm-linux-gnueabihf/libs/libkoreader-cre.so cre.cpp  \
        -lcrengine 
cc1plus: error: unrecognized -mtune target: generic
cc1plus: note: valid arguments are: arm8 arm810 strongarm strongarm110 fa526 fa626 arm7tdmi arm7tdmi-s arm710t arm720t arm740t arm9 arm9tdmi arm920t arm920 arm922t arm940t ep9312 arm10tdmi arm1020t arm9e arm946e-s arm966e-s arm968e-s arm10e arm1020e arm1022e xscale iwmmxt iwmmxt2 fa606te fa626te fmp626 fa726te arm926ej-s arm1026ej-s arm1136j-s arm1136jf-s arm1176jz-s arm1176jzf-s mpcorenovfp mpcore arm1156t2-s arm1156t2f-s cortex-m1 cortex-m0 cortex-m0plus cortex-m1.small-multiply cortex-m0.small-multiply cortex-m0plus.small-multiply generic-armv7-a cortex-a5 cortex-a7 cortex-a8 cortex-a9 cortex-a12 cortex-a15 cortex-a17 cortex-r4 cortex-r4f cortex-r5 cortex-r7 cortex-r8 cortex-m7 cortex-m4 cortex-m3 marvell-pj4 cortex-a15.cortex-a7 cortex-a17.cortex-a7 cortex-a32 cortex-a35 cortex-a53 cortex-a57 cortex-a72 cortex-a73 exynos-m1 xgene1 cortex-a57.cortex-a53 cortex-a72.cortex-a53 cortex-a73.cortex-a35 cortex-a73.cortex-a53 cortex-a55 cortex-a75 cortex-a76 cortex-a76ae cortex-a77 neoverse-n1 cortex-a75.cortex-a55 cortex-a76.cortex-a55 neoverse-v1 neoverse-n2 cortex-m23 cortex-m33 cortex-m35p cortex-m55 cortex-r52 native
make[1]: *** [Makefile:149: build/arm-linux-gnueabihf/libs/libkoreader-cre.so] Error 1
make[1]: Leaving directory '/data/src/koreader/koreader/base'
```

#### Steps to reproduce
run ./kodev release debian

release debian-armhf works

##### full compile log
[deb-compile.log](https://github.com/koreader/koreader/files/9595344/deb-compile.log)

**Comment by NiLuJe** [comment]
> release debian-armhf works

That's because that's how you're supposed to build it, AFAIK ;).

**Comment by pazos** [comment]
Yup.

I didn't know it was possible to use a cross-compilation target without cross-compiling (ie: host and target are the same arch).

Anyhow: removing a few lines would avoid that error: https://github.com/koreader/koreader-base/blob/master/Makefile.defs#L480-L482. This doesn't apply to cross compilation target, which are `DEBIAN_CROSS`

Sadly these lines are useful for x86/x86_64 because without them the package would rely on build computer's intrinsics, enabling avx2/avx512 and making the program crash on legacy cpu's .

I think I've tried to build the emulator natively on arm64 successfully but I'm not sure. Maybe `-mtune generic` is valid for armv8a?

**Comment by NiLuJe** [comment]
> Maybe `-mtune generic` is valid for armv8a

Nah, recent (-ish) versions of GCC pulled the generic `generic` from the arch target, it always needs to be generic-*arch* now.

**Comment by pazos** [comment]
@NiLuJe: do you have any wizard recipes to include `-mtune generic` only on x86 hosts?

We can apply it to `appimage`, `debian` and `macos` targets

**Comment by NiLuJe** [comment]
Check `$CC -dumpmachine` for a tuple starting with x86?

(Don't we already have something like that somewhere?).

**Comment by akemnade** [comment]
hmm, why do we need to add a -mtune=generic and not every deb package does need to do that?

**Comment by NiLuJe** [comment]
Because we're not running in a controlled build environment (where you'd instead use a TC *built* to default to your minimal arch level).

**Comment by akemnade** [comment]
Well, on a debian system, the compiler would default the minimal build arch specified by that debian arch. And for example systemd is compiled without -mtune=generic. So if that would be harmful or worsen things, it does not matter if any application could run or not. 
So why does e.g. systemd not need the flag but koreader does need it? Maybe the answer is simple, perhaps I am just overlooking something simple.

**Comment by NiLuJe** [comment]
IIRC, the Debian packages are in a bit of a weird spot in that they may be used for potentially *more* stuff than purely Debian-land, and may not necessarily be built by Debian TCs, may not necessarily be run on Debian at all, and *need* to be able to run on plain hardware.

As such, going over the TC and enforcing -mtune=generic is the safest way to ensure this behaves across all potential envs.

**Comment by NiLuJe** [comment]
FWIW, c.f., `gcc -Q -v <a random C program>` to see what your compiler actually defaults to.

e.g.,

`COLLECT_GCC_OPTIONS='-Q' '-v' '-mtune=generic' '-march=x86-64' '-dumpdir' 'a-'`

**Comment by akemnade** [comment]
Well, ok, if debian packages are meant that way, than I could understand the -mtune=generic argument. But then the question is whether the upcoming linux-generic package should be used for that cases, and maybe  the debian package then really limited more on debian land. Maybe even build using the dpkg-buildpackage/pbuilder/cowbuilder and friends way.

**Comment by pazos** [comment]
@akemnade: Patches welcome that implement @NiLuJe suggestion in https://github.com/koreader/koreader/issues/9530#issuecomment-1250407684

>  maybe the debian package then really limited more on debian land. Maybe even build using the dpkg-buildpackage/pbuilder/cowbuilder and friends way.

Downstream maintainers can handle the build like they want. We don't care about that. Also we don't want to waste time redoing a build system that works already.

**Comment by NiLuJe** [comment]
Note that I'm not even sure the suggestion is a good idea, for the simple reason that there is no longer a "generic" ARM target.

What *might* make sense instead is a new dedicated target that just does -march=native and assumes the fact that it is 100% dedicated to the build machine.

**Comment by pazos** [comment]
> Note that I'm not even sure the suggestion is a good idea, for the simple reason that there is no longer a "generic" ARM target (and a target that may or may not build for a specific target depending on its triplet would be potentially confusing as all hell if there's ever an issue with the binaries it produces), and we already have specific targets for all the arm variants that matter.
> 

All our arm targets are made on x86_64 :p. This is about building the app for the same machine where the machine is not x86.

> What _might_ make sense instead is a new dedicated target that just does -march=native for a package that would be 100% dedicated to the build machine.

Maybe we can do both at once? 

See https://github.com/koreader/koreader-base/pull/1504

an env var like `BUILDTYPE=generic` | `BUILDTYPE=native` and when no env is set do nothing to enforce specific compiler flags?

**Comment by akemnade** [comment]
> > release debian-armhf works
> 
> That's because that's how you're supposed to build it, AFAIK ;).

I am expecting that I can trust the usage texts of programs:

 ./kodev release 

usage: release <OPTIONS> <TARGET>
[...]
    debian          Debian package for current arch
[...]

So I expect to be able to use debian here for building armhf on armhf. But I do not have a strong opinion here.  Patching the help  text would also solve that.

**Comment by akemnade** [comment]
I am studying koreader/koreader-base#1504 and will probably continue there.
